exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/PcapRecordReader.java

```diff
@@ -0,0 +1,295 @@
+import org.apache.drill.common.expression.SchemaPath;
+import org.apache.drill.common.types.TypeProtos;
+import org.apache.drill.common.types.TypeProtos.MajorType;
+import org.apache.drill.common.types.TypeProtos.MinorType;
+import org.apache.drill.common.types.Types;
+import org.apache.drill.exec.exception.SchemaChangeException;
+import org.apache.drill.exec.expr.TypeHelper;
+import org.apache.drill.exec.ops.OperatorContext;
+import org.apache.drill.exec.physical.impl.OutputMutator;
+import org.apache.drill.exec.record.MaterializedField;
+import org.apache.drill.exec.store.AbstractRecordReader;
+import org.apache.drill.exec.store.pcap.decoder.Packet;
+import org.apache.drill.exec.store.pcap.decoder.PacketDecoder;
+import org.apache.drill.exec.store.pcap.dto.ColumnDto;
+import org.apache.drill.exec.store.pcap.schema.PcapTypes;
+import org.apache.drill.exec.store.pcap.schema.Schema;
+import org.apache.drill.exec.vector.NullableBigIntVector;
+import org.apache.drill.exec.vector.NullableIntVector;
+import org.apache.drill.exec.vector.NullableTimeStampVector;
+import org.apache.drill.exec.vector.NullableVarCharVector;
+import org.apache.drill.exec.vector.ValueVector;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.drill.exec.store.pcap.Utils.parseBytesToASCII;
+
+public class PcapRecordReader extends AbstractRecordReader {
+
+  private OutputMutator output;
+
+  private final PacketDecoder decoder;
+  private ImmutableList<ProjectedColumnInfo> projectedCols;
+
+  private byte[] buffer = new byte[100000];
+  private int offset = 0;
+  private InputStream in;
+  private int validBytes;
+
+  private static final Map<PcapTypes, MinorType> TYPES;
+
+  private static class ProjectedColumnInfo {
+    ValueVector vv;
+    ColumnDto pcapColumn;
+  }
+
+  static {
+    TYPES = ImmutableMap.<PcapTypes, TypeProtos.MinorType>builder()
+        .put(PcapTypes.STRING, MinorType.VARCHAR)
+        .put(PcapTypes.INTEGER, MinorType.INT)
+        .put(PcapTypes.LONG, MinorType.BIGINT)
+        .put(PcapTypes.TIMESTAMP, MinorType.TIMESTAMP)
+        .build();
+  }
+
+  public PcapRecordReader(final String inputPath,
+                          final List<SchemaPath> projectedColumns) {
+    try {
+      this.in = new FileInputStream(inputPath);
+      this.decoder = getPacketDecoder();
+      validBytes = in.read(buffer);
+    } catch (IOException e) {
+      throw new RuntimeException("File " + inputPath + " not Found");
+    }
+    setColumns(projectedColumns);
+  }
+
+  @Override
+  public void setup(final OperatorContext context, final OutputMutator output) throws ExecutionSetupException {
+    this.output = output;
+  }
+
+  @Override
+  public int next() {
+    projectedCols = getProjectedColsIfItNull();
+    try {
+      return parsePcapFilesAndPutItToTable();
```

Comment:
Drill has certain protocols that are not entirely obvious, but that are needed here. Each call to `next()` produces a batch of rows. Drill requires that the batch contain no more than 64K rows. (Bad things happen if the batch has more.) So, the trick for a record reader is to read up to some maximum number of records into each batch, return, and wait to be called again for the next batch.

Depending on the size of the data, 64K may be to large and 8K or some other number may be better. Try to avoid causing any vector to be larger than 16 MB. (We're working on a project to enforce this, and the result will make it much easier to write a record reader. Too bad your project is just a bit too early to benefit.)

If I read this code correctly, each call to `next()` returns just one record, until EOF at which time it returns 0.

You'll find that the rest of Drill will be *very* slow with a batch size of 1 record. So, perhaps put this into a loop to read 8K or so rows into a single batch, then return the total record count. You should see a nice execution speed-up.